server/src/test/java/org/candlepin/async/tasks/EntitleByProductsJobTest.java

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright (c) 2009 - 2019 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+
+package org.candlepin.async.tasks;
+
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import org.candlepin.async.JobConfig;
+import org.candlepin.async.JobConfigValidationException;
+import org.candlepin.async.JobExecutionContext;
+import org.candlepin.controller.Entitler;
+import org.candlepin.model.Consumer;
+import org.candlepin.model.ConsumerType;
+import org.candlepin.model.Entitlement;
+import org.candlepin.model.Owner;
+import org.candlepin.pinsetter.tasks.BaseJobTest;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+
+public class EntitleByProductsJobTest extends BaseJobTest {
+
+    private Consumer consumer;
+    private Owner owner;
+    private String consumerUuid;
+    private Entitler entitler;
+
+    @BeforeEach
+    public void init() {
+        super.init();
+        consumerUuid = "49bd6a8f-e9f8-40cc-b8d7-86cafd687a0e";
+
+        final ConsumerType ctype = new ConsumerType("system");
+        ctype.setId("test-ctype");
+        owner = new Owner("test-owner");
+        owner.setId("test-owner-id");
+        consumer = new Consumer("Test Consumer", "test-consumer", owner, ctype);
+        consumer.setUuid(consumerUuid);
+        entitler = mock(Entitler.class);
+    }
+
+    @Test
+    void allValuesPresent() {
+        final String[] pids = { "pid1", "pid2", "pid3" };
+        final List<String> fromPools = Collections.singletonList("pool_id1");
+
+        final JobConfig config = EntitleByProductsJob.createConfig()
+            .setOwner(owner)
+            .setConsumer(consumer)
+            .setProductIds(pids)
+            .setEntitleDate(new Date())
+            .setFromPools(fromPools);
+
+        assertDoesNotThrow(config::validate);
```

Comment:
These tests are kind of okay, but testing for correctness based on the absence (or non-absence) of a value isn't complete. Also, this particular implementation relies on validate being correct, which should be tested on its own.

A better way to approach this would be to test that each method is setting the appropriate fields in isolation.

Reply:
I am not testing those setters directly. In these tests I am using them to test validation. Whether values are retrievable is tested below when job tries to access those values.